lib/src/compile.ts

```diff
@@ -0,0 +1,156 @@
+  source: string;
+  sourceMap?: (sourceMap: RawSourceMap) => void;
+  url?: URL;
+}): Promise<string> {
+  // TODO(awjin): Create logger, importer, function registries.
+
+  const request = newCompileStringRequest({
+    source: options.source,
+    sourceMap: !!options.sourceMap,
+    url: options.url,
+  });
+
+  const response = await compileRequest(request);
+  if (options.sourceMap) {
+    options.sourceMap(response.sourceMap!);
+  }
+  return response.css;
+}
+
+// Creates a request for compiling a file.
+function newCompileRequest(options: {
+  path: string;
+  sourceMap: boolean;
+}): InboundMessage.CompileRequest {
+  // TODO(awjin): Populate request with importer/function IDs.
+
+  const request = new InboundMessage.CompileRequest();
+  request.setPath(options.path);
+  request.setSourceMap(options.sourceMap);
+
+  return request;
+}
+
+// Creates a request for compiling a string.
+function newCompileStringRequest(options: {
+  source: string;
+  sourceMap: boolean;
+  url?: URL;
+}): InboundMessage.CompileRequest {
+  // TODO(awjin): Populate request with importer/function IDs.
+
+  const input = new InboundMessage.CompileRequest.StringInput();
+  input.setSource(options.source);
+  if (options.url) input.setUrl(options.url.toString());
+
+  const request = new InboundMessage.CompileRequest();
+  request.setString(input);
+  request.setSourceMap(options.sourceMap);
+
+  return request;
+}
+
+// Spins up a compiler, then sends it a compile request. Returns a promise that
+// resolves with the CompileResult. Throws if there were any protocol or
+// compilation errors. Shuts down the compiler after compilation.
+async function compileRequest(
+  request: InboundMessage.CompileRequest
+): Promise<{
+  css: string;
+  sourceMap?: RawSourceMap;
+}> {
+  // TODO(awjin):
+  // - Subscribe logger to dispatcher's log events.
+  // - Pass import and function registries' handler functions to dispatcher.
+
+  const embeddedCompiler = new EmbeddedCompiler();
+  const packetTransformer = new PacketTransformer(
+    embeddedCompiler.stdout$,
+    buffer => embeddedCompiler.writeStdin(buffer)
+  );
+  const messageTransformer = new MessageTransformer(
+    packetTransformer.outboundProtobufs$,
+    packet => packetTransformer.writeInboundProtobuf(packet)
+  );
+  const dispatcher = new Dispatcher(
+    messageTransformer.outboundMessages$,
+    message => messageTransformer.writeInboundMessage(message),
+    {
+      handleImportRequest: () => {
+        throw Error('Custom importers not yet implemented.');
+      },
+      handleFileImportRequest: () => {
+        throw Error('Custom file importers not yet implemented.');
+      },
+      handleCanonicalizeRequest: () => {
+        throw Error('Canonicalize not yet implemented.');
+      },
+      handleFunctionCallRequest: () => {
+        throw Error('Custom functions not yet implemented.');
+      },
+    }
+  );
+
+  dispatcher.error$.subscribe(error => {
+    throw error;
```

Comment:
I believe this will cause the error to be passed to the top-level error handler. Rather than that, we probably want to pipe it into the returned `Promise`.

Reply:
Since `compileRequest` is an `async` function, it implicitly wraps everything in a promise. This `throw` then fails that promise

Reply:
Nope, I'm wrong. The subscription could fail after the promise resolves. Will fix

Reply:
Although, line 139 *should* close the subscription before this promise returns. Maybe still worth unsubscribing explicitly?

Reply:
Yeah, an `async` function doesn't automatically collect all errors in its lexical scope—just those that are emitted by `await`ed Promises. You'll need to essentially "race" this error against the `sendCompileRequest()` promise, so that if `dispatcher.error$` emits an error before the promise completes `compileRequest()` forwards that error but if it doesn't the stream gets unsubscribed.

Reply:
Sigh I was massively overthinking this. The dispatcher already wraps up the error into the promise that it returns. I just have to handle it properly with a `finally` clause.